hibernate-core/src/main/java/org/hibernate/dialect/JsonHelper.java

```diff
@@ -381,996 +448,52 @@ public static <X> X arrayFromString(
-						case VALUE_QUOTE:
-							// In the value it's fine
-							break;
-						case VALUE_END:
-							// At this point, we are done
-							return i + 1;
-						default:
-							throw syntaxError( string, s, i );
-					}
-					break;
-				default:
-					switch ( s ) {
-						case KEY_QUOTE:
-						case VALUE_QUOTE:
-							// In keys and values, all chars are fine
-							break;
-						case VALUE_START:
-							// Skip whitespace
-							if ( Character.isWhitespace( c ) ) {
-								break;
-							}
-							// Here we also allow certain literals
-							final int endIdx = consumeLiteral(
-									string,
-									i,
-									values,
-									embeddableMappingType.getJdbcValueSelectable( selectableIndex ).getJdbcMapping(),
-									selectableIndex,
-									returnEmbeddable,
-									options
-							);
-							if ( endIdx != -1 ) {
-								i = endIdx;
-								s = State.VALUE_END;
-								selectableIndex = -1;
-								start = -1;
-								break;
-							}
-							throw syntaxError( string, s, i );
-						case KEY_START:
-						case KEY_END:
-						case VALUE_END:
-							// Only whitespace is allowed here
-							if ( Character.isWhitespace( c ) ) {
-								break;
-							}
-						default:
-							throw syntaxError( string, s, i );
-					}
-					break;
-			}
-		}
-
-		throw new IllegalArgumentException( "JSON not properly formed: " + string.subSequence( start, end ) );
-	}
+		JsonValueJDBCTypeAdapter adapter = JsonValueJDBCTypeAdapterFactory.getAdapter(reader,false);
 
-	private static int fromArrayString(
-			String string,
-			boolean returnEmbeddable,
-			WrapperOptions options,
-			int begin,
-			CustomArrayList arrayList,
-			BasicType<?> elementType) throws SQLException {
-		return fromArrayString(
-				string,
-				returnEmbeddable,
-				options,
-				begin,
-				arrayList,
-				elementType.getMappedJavaType(),
-				elementType.getJdbcJavaType(),
-				elementType.getJdbcType()
-		);
-	}
-
-	private static int fromArrayString(
-			String string,
-			boolean returnEmbeddable,
-			WrapperOptions options,
-			int begin,
-			CustomArrayList arrayList,
-			JavaType<?> javaType,
-			JavaType<?> jdbcJavaType,
-			JdbcType jdbcType) throws SQLException {
-		if ( string.length() == begin + 2 ) {
-			return begin + 2;
-		}
-		boolean hasEscape = false;
-		assert string.charAt( begin ) == '[';
-		int start = begin + 1;
-		State s = State.VALUE_START;
-		// The following parsing logic assumes JSON is well-formed,
-		// but for the sake of the Java compiler's flow analysis
-		// and hopefully also for a better understanding, contains throws for some syntax errors
-		for ( int i = start; i < string.length(); i++ ) {
-			final char c = string.charAt( i );
-			switch ( c ) {
-				case '\\':
-					assert s == State.VALUE_QUOTE;
-					hasEscape = true;
-					i++;
+		assert reader.hasNext():"Invalid array string";
+		assert reader.next() == JsonDocumentItemType.ARRAY_START:"Invalid start of array";
```

Comment:
Let's add proper checks:

```suggestion
		final JsonDocumentItemType event;
		if ( !reader.hasNext() || ( event = reader.next() ) == NULL_VALUE ) {
			return null;
		}
		if ( event != ARRAY_START ) {
			throw new IllegalArgumentException("Malformed JSON. Expected array but got: " + event);
		}
```